templates/e2e-dspy-workflow/src/utils.py

```diff
@@ -0,0 +1,115 @@
+import dspy
+import dsp
+import os
+
+# -------------------------------
+
+class IntentClassification(dspy.Signature):
+    """As a part of a banking issue traiging system, classify the intent of a natural language query into one of the 25 labels.
+    The intent should exactly match one of the following:
+    ['card_payment_fee_charged', 'direct_debit_payment_not_recognised', 'balance_not_updated_after_cheque_or_cash_deposit', 'wrong_amount_of_cash_received', 'cash_withdrawal_charge', 'transaction_charged_twice', 'declined_cash_withdrawal', 'transfer_fee_charged', 'balance_not_updated_after_bank_transfer', 'transfer_not_received_by_recipient', 'request_refund', 'card_payment_not_recognised', 'card_payment_wrong_exchange_rate', 'extra_charge_on_statement', 'wrong_exchange_rate_for_cash_withdrawal', 'refund_not_showing_up', 'reverted_card_payment', 'cash_withdrawal_not_recognised', 'activate_my_card', 'pending_card_payment', 'cancel_transfer', 'beneficiary_not_allowed', 'card_arrival', 'declined_card_payment', 'pending_top_up']
+    """
+
+    intent = dspy.InputField(desc="Intent of the query")
+    label = dspy.OutputField(desc="Type of the intent; Should just be one of the 25 labels with no other text")
+
+class IntentClassificationModule(dspy.Module):
+    def __init__(self, labels_in_use):
+        self.intent_classifier = dspy.ChainOfThought(IntentClassification)
+        self.valid_labels = set(labels_in_use)
+
+    def forward(self, text):
+        prediction = self.intent_classifier(intent=text)
+        sanitized_prediction = dspy.Prediction(label=prediction.label.lower().strip().replace(" ", "_"), reasoning=prediction.reasoning)
+        return sanitized_prediction
```

Comment:
Done this way because I dont know any easy ways to import modules from the main notebook. Could always pass around the class references in the notebook.